### PR TITLE
[WIP] Task 4706: Add Data-Iterators tests

### PR DESCRIPTION
**Pull request by sergei820**

### CODE REVIEW

File: jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/complex/DataIteratorsTests.java [hunk 1]
```diff
@@ -99,4 +107,37 @@ public void singleSelectDataIteratorTest() {
...
+        loadingTextDataIterator.has().text(containsString("Items are loading, please, wait a little bit..."));
+        //The default text is "Loading items...".
+        // Here we also check the parameter loading-text="..." that allows to set custom text value for the parameter.
+    }
+
+    @Test(description = "Test checks data iterator parameter : Group by")
+    public void groupByDataIteratorTest() {
+        groupedByDataIterator.show();
+        groupedByDataIterator.has().size(greaterThan(1));
+    }
```
Comment (MayaElf) [inline jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/complex/DataIteratorsTests.java:142]: Please, try to find a way how to check the main functionality here: 'groupig' items by chosen parameter
`desserts: [
        {
          name: 'Frozen Yogurt',
          calories: 159,
          fat: 6.0,
          carbs: 24,
          protein: 4.0,
          sodium: 87,
          calcium: '14%',
          iron: '1%',
        },
        {
          name: 'Ice cream sandwich',
          calories: 237,
          fat: 9.0,
          carbs: 37,
          protein: 4.3,
          sodium: 129,
          calcium: '8%',
          iron: '1%',
        },
        {
          name: 'Eclair',
          calories: 262,
          fat: 16.0,
          carbs: 23,
          protein: 6.0,
          sodium: 337,
          calcium: '6%',
          iron: '7%',
        },
        {
          name: 'Cupcake',
          calories: 305,
          fat: 3.7,
          carbs: 67,
          protein: 4.3,
          sodium: 413,
          calcium: '3%',
          iron: '8%',
        },
      ],`

**Review by MayaElf — CHANGES_REQUESTED** [review]

File: jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/complex/DataIteratorsTests.java (hunk 1 above)
Comment (sergei820) [reply jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/complex/DataIteratorsTests.java:151]: Implemented DataProvider solution in a little bit different way, please check

**Comment by pnatashap** [comment]
Check the task description once again, it was changed. 
Total review for this element is needed

**Review by MayaElf — APPROVED** [review]

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/asserts/CardAssert.java [hunk 2]
```diff
@@ -31,4 +31,5 @@ public CardAssert assertNumberOfButtonsOnCard(int number) {
         jdiAssert(element().getButtons().size() == number, Matchers.is(true));
         return this;
     }
+
```
Comment (pnatashap) [inline jdi-light-angular/src/main/java/com/epam/jdi/light/angular/asserts/CardAssert.java:34]: revert this change, it is another project

**Review by pnatashap — CHANGES_REQUESTED** [review]

**Comment by pnatashap** [comment]
@sergei820 please fix review comments

File: jdi-light-vuetify-tests/src/test/resources/test.properties [hunk 3]
```diff
@@ -15,6 +15,7 @@ log.level=INFO
 
 #page.load.strategy=normal
 #browser.size=MAXIMIZE | 1024x762
+browser.size=1024x762
```
[the remark this answers is not in the document]
Comment (sergei820) [reply jdi-light-vuetify-tests/src/test/resources/test.properties]: Last commit doesn't include any changes

File: jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/complex/DataIteratorsTests.java [hunk 4]
```diff
@@ -99,4 +109,42 @@ public void singleSelectDataIteratorTest() {
         singleSelectTableIterator.is().unchecked(0);
         singleSelectTableIterator.is().checked(1);
     }
+
+    @Test(description = "Test checks data iterator parameter : No data text")
+    public void noDataTextDataIteratiorTest() {
+        noDataTextDataIterator.show();
+        noDataTextDataIterator.has().text(containsString("Sorry, there's no any items data..."));
+        //The default text is "No data available".
+        //Here we also check the parameter no-data-text="..." that allows to set custom text value for the parameter.
```
[the remark this answers is not in the document]
Comment (sergei820) [reply jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/complex/DataIteratorsTests.java]: Done

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/complex/Card.java [hunk 5]
```diff
@@ -137,4 +138,50 @@ public void click() {
     public CardAssert is() {
         return new CardAssert().set(this);
     }
+
+    //Functionality for Cards (with Switch) in DataIterators
+
+    private UIElement expander() {
```
[the remark this answers is not in the document]
Comment (sergei820) [reply jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/complex/Card.java:144]: Not sure if I understand you correctly, there is this element on the test site
image

So, should I do anything with this element?

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/asserts/tables/DataIteratorAssert.java [hunk 6]
```diff
@@ -101,4 +102,10 @@ public DataIteratorAssert footerWithDarkTheme() {
         jdiAssert(element().footerTheme(), Matchers.is("theme--dark"), "Element's footer has light theme");
```
[the remark this answers is not in the document]
Comment (sergei820) [reply jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/asserts/tables/DataIteratorAssert.java]: Done

File: jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/complex/DataIteratorsTests.java [hunk 7]
```diff
@@ -99,4 +109,42 @@ public void singleSelectDataIteratorTest() {
...
+        loadingTextDataIterator.show();
+        loadingTextDataIterator.has().text(containsString("Items are loading, please, wait a little bit..."));
+        //The default text is "Loading items...".
+        // Here we also check the parameter loading-text="..." that allows to set custom text value for the parameter.
+    }
+
+    @Test(description = "Test checks data iterator parameter : Group by",
+    dataProvider = "groupByDataIteratorTestData", dataProviderClass = DataIteratorDataProvider.class)
+    public void groupByDataIteratorTest(String groupingParameter, String parameterValue, int expectedElementsQuantity) {
+        WebList groupedByDataIterator = $$(groupedDataIteratorLocator
```
[the remark this answers is not in the document]
Comment (sergei820) [reply jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/complex/DataIteratorsTests.java]: I need a dynamic locator here, to generate it each time with data from dataProvider. Moved the functionality from tests layer, please, take a look.

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/asserts/tables/DataIteratorAssert.java [hunk 8]
```diff
@@ -101,4 +102,10 @@ public DataIteratorAssert footerWithDarkTheme() {
...
         return this;
     }
+
+    @JDIAction("Assert that '{name}' text {0}")
+    public DataIteratorAssert text(Matcher<String> condition) {
```
[the remark this answers is not in the document]
Comment (sergei820) [reply jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/asserts/tables/DataIteratorAssert.java:76]: Please, tell me, if I should change something. 
It's not huge:
image

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/asserts/CardAssert.java (hunk 2 above)
Comment (sergei820) [reply jdi-light-angular/src/main/java/com/epam/jdi/light/angular/asserts/CardAssert.java:34]: Please, specify, what exactly should I revert? The whole Card functionality should be returned to DataIterator?

**Comment by sergei820** [comment]
@pnatashap, please, review my fixes and questions

**Comment by pnatashap** [comment]
@sergei820 there is a special button to mark PR as ready to review again. Now I see that PR is still in progress of changes

File: jdi-light-angular/src/main/java/com/epam/jdi/light/angular/asserts/CardAssert.java (hunk 2 above)
Comment (pnatashap) [reply jdi-light-angular/src/main/java/com/epam/jdi/light/angular/asserts/CardAssert.java:34]: This is another project jdi-light-angular, we are working ONLY in vuetify projects and do not use that one

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/complex/tables/DataIterator.java [hunk 9]
```diff
@@ -1,94 +1,34 @@
...
 
     public WebList dataIteratorElements() {
         return finds("[class^='col']");
     }
 
-    private UIElement expander(int colNum) {
-        return dataIteratorElements().get(colNum).find("[class*='selection']");
-    }
-
     public Integer getColumnsValue() {
```
Comment (pnatashap) [inline jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/complex/tables/DataIterator.java:26]: причем тут columns? нет у нас колонок

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/complex/tables/DataIterator.java [hunk 10]
```diff
@@ -1,94 +1,34 @@
...
-
-    @JDIAction("Get '{name}' column title")
-    public String getColumnTitle(int colNum) {
-        return dataIteratorElements().get(colNum).find(TITLE).getText();
-    }
-
-    @JDIAction("Is '{name}' column empty")
-    public boolean columnIsEmpty(int colNum) {
-        return getColumn(colNum).isEmpty();
+    public Card dataiteratorCardByNumber(int cardNumber) {
```
Comment (pnatashap) [inline jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/complex/tables/DataIterator.java:30]: в data iterator могут быть не только Card внутри, любые элементы могут быть

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/complex/tables/DataIterator.java [hunk 11]
```diff
@@ -1,94 +1,34 @@
...
+        return dataIteratorElements().get(cardNumber).with(Card.class);
     }
 
     @JDIAction("Get '{name}' iterator")
```
Comment (pnatashap) [inline jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/complex/tables/DataIterator.java:34]: нет у нас таблицы

File: jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/complex/tables/DataIterator.java [hunk 12]
```diff
@@ -149,4 +88,10 @@ public DataIteratorAssert assertThat() {
     public DataIteratorAssert has() {
         return is();
     }
+
+    public static WebList collectElementsGroupedByParameterWithValue(String groupingParameter, String parameterValue) {
```
Comment (pnatashap) [inline jdi-light-vuetify/src/main/java/com/epam/jdi/light/vuetify/elements/complex/tables/DataIterator.java:92]: а какое отношение этот метод имеет к этому классу? это кастомная реализация, которая к общему классу не имеет отношения

File: jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/complex/DatePickersTests.java [hunk 13]
```diff
@@ -53,7 +60,6 @@
 import static org.hamcrest.Matchers.containsInAnyOrder;
 import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.is;
```
Comment (pnatashap) [inline jdi-light-vuetify-tests/src/test/java/io/github/epam/vuetify/tests/complex/DatePickersTests.java:62]: откати эти изменения не создавай проблем коллегам - у них конфликт слияния будет из-за этого

**Review by pnatashap — CHANGES_REQUESTED** [review]
много лишнего осталось

**Comment by pnatashap** [comment]
PR is not good, but merging to fix in the main brunch
